app/abci/handlers.go

```diff
@@ -0,0 +1,222 @@
+)
+
+type Handlers struct {
+	batchingKeeper        BatchingKeeper
+	pubKeyKeeper          PubKeyKeeper
+	stakingKeeper         StakingKeeper
+	validatorAddressCodec addresscodec.Codec
+	signer                utils.SEDASigner
+	logger                log.Logger
+}
+
+func NewHandlers(
+	bk BatchingKeeper,
+	pkk PubKeyKeeper,
+	sk StakingKeeper,
+	vac addresscodec.Codec,
+	logger log.Logger,
+) *Handlers {
+	return &Handlers{
+		batchingKeeper:        bk,
+		pubKeyKeeper:          pkk,
+		stakingKeeper:         sk,
+		validatorAddressCodec: vac,
+		logger:                logger,
+	}
+}
+
+func (h *Handlers) SetSEDASigner(signer utils.SEDASigner) {
+	h.signer = signer
+}
+
+// ExtendVoteHandler handles the ExtendVote ABCI to sign a batch created
+// from the previous block.
+func (h *Handlers) ExtendVoteHandler() sdk.ExtendVoteHandler {
+	return func(ctx sdk.Context, _ *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+		h.logger.Debug("start extend vote handler")
+
+		// Sign the batch created from the last block's end blocker.
+		batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()-1)
+		if err != nil {
+			return nil, err
+		}
+		signature, err := h.signer.Sign(batch.BatchId, utils.Secp256k1)
+		if err != nil {
+			return nil, err
+		}
+
+		h.logger.Debug("submitting batch signature", "signature", signature)
+		return &abci.ResponseExtendVote{VoteExtension: signature}, nil
+	}
+}
+
+// VerifyVoteExtensionHandler handles the VerifyVoteExtension ABCI to
+// verify the batch signature included in the pre-commit vote against
+// the public key registered in the pubkey module.
+func (h *Handlers) VerifyVoteExtensionHandler() sdk.VerifyVoteExtensionHandler {
+	return func(ctx sdk.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error) {
+		h.logger.Debug("start verify vote extension handler", "request", req)
+
+		batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()-1)
+		if err != nil {
+			return nil, err
+		}
+		ok, err := h.verifyBatchSignature(ctx, batch.BatchId, req.VoteExtension[:64], req.ValidatorAddress, utils.Secp256k1)
+		if !ok || err != nil {
+			return nil, err
+		}
+
+		h.logger.Debug("successfully verified signature", "request", req.ValidatorAddress, "height", req.Height)
+		return &abci.ResponseVerifyVoteExtension{Status: abci.ResponseVerifyVoteExtension_ACCEPT}, nil
+	}
+}
+
+// PrepareProposalHandler handles the PrepareProposal ABCI to inject
+// a canonical set of vote extensions in the proposal.
+func (h *Handlers) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		h.logger.Debug("start prepare proposal handler")
+
+		err := baseapp.ValidateVoteExtensions(ctx, h.stakingKeeper, req.Height, ctx.ChainID(), req.LocalLastCommit)
+		if err != nil {
+			return nil, err
+		}
+
+		proposalTxs := req.Txs
```

Comment:
Should we consider max block size?

Reply:
Now the handlers wrap the default proposal handlers and run them after the vote injection logic.